python/lsst/daf/butler/core/serverModels.py

```diff
@@ -82,33 +93,62 @@ def from_expression(cls, expression: Any) -> ExpressionQueryParameter:
             return cls()
 
         expressions = iterable(expression)
-        params: Dict[str, List[str]] = {}
+        params: Dict[str, List[str]] = {"glob": [], "regex": []}
         for expression in expressions:
             if expression is ...:
                 # This matches everything
                 return cls()
 
             if isinstance(expression, re.Pattern):
-                if (k := "regex") not in params:
-                    params[k] = []
                 params["regex"].append(expression.pattern)
-            else:
-                if (k := "glob") not in params:
-                    params[k] = []
+            elif isinstance(expression, str):
                 params["glob"].append(expression)
+            elif hasattr(expression, "name"):
+                params["glob"].append(expression.name)
+            else:
+                raise ValueError(f"Unrecognized type given to expression: {expression!r}")
+
+        # Clean out empty dicts.
+        for k in list(params):
+            if not params[k]:
+                del params[k]
+
         return cls(**params)
 
 
+class DatasetsQueryParameter(ExpressionQueryParameter):
+    """Represents a specification for a dataset expression query.
+
+    This differs from the standard expression query in that an empty
+    expression will return `None` rather than ``...``.
+    """
+
+    _allow_ellipsis: ClassVar[bool] = False
+
+
 class QueryDatasetsModel(BaseModel):
     """Information needed for a registry dataset query."""
 
     datasetType: ExpressionQueryParameter
     collections: Optional[ExpressionQueryParameter] = None
     dimensions: Optional[List[str]] = None
-    dataId: Optional[DataId] = None
+    dataId: Optional[SerializedDataCoordinate] = None
     where: Optional[str] = None
     findFirst: bool = False
     components: Optional[bool] = None
-    bind: Optional[DataId] = None
+    bind: Optional[SimpleDataId] = None
+    check: bool = True
+    keyword_args: Optional[SimpleDataId] = None  # mypy refuses to allow kwargs in model
+
+
+class QueryDimensionRecordsModel(BaseModel):
+    """Information needed to query the dimension records."""
+
+    dataId: Optional[SerializedDataCoordinate] = None
+    datasets: Optional[DatasetsQueryParameter] = None
+    collections: Optional[ExpressionQueryParameter] = None
+    where: Optional[str] = None
+    components: Optional[bool] = None
+    bind: Optional[SimpleDataId] = None
```

Comment:
This is not a data ID; the mapping is from `str` to any type that a `DimensionRecord` attribute can have (so at least also `bool`, `float`, and `Timespan`, in addition to `int` and `str`).